test/integ/routers/alpha-router/alpha-router.integration.test.ts

```diff
@@ -3558,29 +3563,74 @@ describe('quote for other networks', () => {
                 }
               );
-              expect(swap).toBeDefined();
-              expect(swap).not.toBeNull();
-              if (swap) {
+              expect(swapWithSimulation).toBeDefined();
+              expect(swapWithSimulation).not.toBeNull();
+
+              // TODO: We need to fix L2 -> L1 calldata posting gas estimate one by one.
+              //       We started by fixing Arbitrum gas estimate https://github.com/Uniswap/smart-order-router/pull/468.
+              //       Before the fix, the non-simulated gas estimate is 131000, meanwhile simulated gas estimate is 6573602.
+              //       After the fix, the non-simulated gas estimate is now 3414207.
+              //       We assert the non-simulated gas estimate is within 50% error margin of simulated gas estimate range, e.g. between 6573602 - 6573602 / 2 and 6573602 + 6573602 / 2
+              if (chain === ChainId.ARBITRUM_ONE) {
+                // Universal Router is not deployed on Gorli.
+                const swapOptions: SwapOptions =
+                  {
+                    type: SwapType.UNIVERSAL_ROUTER,
+                    recipient: WHALES(tokenIn),
+                    slippageTolerance: SLIPPAGE,
+                    deadlineOrPreviousBlockhash: parseDeadline(360),
+                  };
+
+                const swap = await alphaRouter.route(
+                  amount,
+                  getQuoteToken(tokenIn, tokenOut, tradeType),
+                  tradeType,
+                  swapOptions,
+                  {
+                    // @ts-ignore[TS7053] - complaining about switch being non exhaustive
+                    ...DEFAULT_ROUTING_CONFIG_BY_CHAIN[chain],
+                    protocols: [Protocol.V3, Protocol.V2],
+                    saveTenderlySimulationIfFailed: true,
+                  }
+                );
+
+                expect(swap).toBeDefined();
+                expect(swap).not.toBeNull();
+
+                const gasEstimateDiff = swapWithSimulation!.estimatedGasUsed.gt(swap!.estimatedGasUsed)
+                  ? swapWithSimulation!.estimatedGasUsed.sub(swap!.estimatedGasUsed)
+                  : swap!.estimatedGasUsed.sub(swapWithSimulation!.estimatedGasUsed);
+
+                // We will rely on Tenderly gas estimate as source of truth against SOR non-simulated gas estimate accuracy.
+                // This is the only reliable and long-term feasible test assertion approach.
+                // For example, in the near future, after EIP-4844, we expect the gas estimate to drop by (3 / 16)
+                // due to gas cost per compressed calldata byte dropping from 16 to 3.
+                // Relying on Tenderly gas estimate is the only way our github CI can auto catch this.
+                const percentDiff = gasEstimateDiff.mul(BigNumber.from(100)).div(swapWithSimulation!.estimatedGasUsed);
+                expect(percentDiff.lte(BigNumber.from(50))).toBe(true);
```

Comment:
this is a little generous but oh well; maybe 30%? you tell me

Reply:
Yea it's kinda unfortunate, that SOR [V3HeuristicGasModelFactory](https://github.com/Uniswap/smart-order-router/blob/main/src/routers/alpha-router/gas-models/v3/v3-heuristic-gas-model.ts) can still be up to 50% smaller than the simulated gas... But yea, it's a lot of improvement what it used to be. It used to return 358000 gas units for a single route single hop before this PR, meanwhile the actual gas estimate can be around 6573602, but now it returns 3414207, for this test.

In terms of the network cost that's showing on https://app.uniswap.org/ though, I think 50% smaller than the actual gas estimate might not be as a big ideal as when it's 10x smaller than the actual gas estimate. Because Arb (L1 + L2) costs are usually between tens of cents and couple dollars in USD. If it's 50% smaller than the actual gas estimate, then https://app.uniswap.org/ will show the network cost in double. But when it's 10x smaller than the actual gas estimate, then https://app.uniswap.org/ will show the network cost in tenfolded. This is when the user experience visibly degrades.